### PR TITLE
[CDAP-18555] Create Kubernetes namespaces

### PR DESCRIPTION
**Pull request by nliu132**
When running in a Kubernetes environment, have namespace admin create/delete a corresponding k8 namespace. If CPU/memory limits are passed in the `NamespaceMeta`, create a k8 namespace resource quota as well.

JIRA: CDAP-18555

Testing:
Used HTTP calls to create (and later delete) a namespace named `Kubernetes_TEST` with k8s namespace `kube-test`, k8s namespace memory limit 1Gi, and k8s namespace CPU limit 1. Corresponding k8 namespace and resource quota were created
```
$ kubectl get namespaces --show-labels
NAME              STATUS   AGE     LABELS
cdap-system       Active   4d9h    control-plane=cdap-operator,name=cdap-system
default           Active   4d10h   <none>
elastic-system    Active   4d9h    app.kubernetes.io/version=1.3.1,control-plane=elastic-operator,name=elastic-system
kube-node-lease   Active   4d10h   <none>
kube-public       Active   4d10h   <none>
kube-system       Active   4d10h   <none>
kube-test         Active   4s      cdap.namespace=Kubernetes_TEST

$ kubectl get resourcequotas --namespace=kube-test --show-labels
NAME                  AGE   REQUEST                                                                                                                            LIMIT                                   LABELS
cdap-resource-quota   7s                                                                                                                                       limits.cpu: 0/1, limits.memory: 0/1Gi   cdap.namespace=Kubernetes_TEST
gke-resource-quotas   7s    count/ingresses.extensions: 0/5k, count/ingresses.networking.k8s.io: 0/5k, count/jobs.batch: 0/10k, pods: 0/5k, services: 0/1500                                           <none>
```

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 1]
```diff
@@ -553,4 +609,27 @@ private void populateLabels(Map<String, String> sparkConfMap) {
...
+      }
+      String kubeNamespace = namespaceList.getItems().get(0).getMetadata().getName();
+
+      // PropagationPolicy is set to background cascading deletion. Kubernetes deletes the owner object immediately and
+      // the controller cleans up the dependent objects in the background.
+      coreV1Api.deleteNamespace(kubeNamespace, null, null, 0, null, "Background", null);
+      LOG.info("Deleted Kubernetes namespace and associated resources for {}", kubeNamespace);
+    } catch (ApiException e) {
+      LOG.error(e.getResponseBody());
+      throw new IOException("Error occurred while deleting Kubernetes namespace.", e);
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: IIRC @albertshau mentioned that ApiException message is unhelpful, so it would be good to include the response body in the exception message (instead of adding a separate error log line)
i.e. something like this -
throw new IOException("Error occurred while deleting Kubernetes namespace, body =  " + e.getResponseBody(), e);

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 2]
```diff
@@ -280,6 +285,57 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+      }
+    } catch (ApiException e) {
+      LOG.error(e.getResponseBody());
+      throw new IOException("Cannot check if Kubernetes namespace already exists", e);
+    }
+
+    V1Namespace namespace = new V1Namespace().apiVersion("v1").kind("Namespace");
+    namespace.setMetadata(new V1ObjectMeta().name(kubeName).putLabelsItem("cdapNamespace", cdapNamespace));
+
+    String kubeCpuLimit = properties.get("kube.limits.cpu");
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Use constants  (ex - KUBE_CPU_LIMIT) wherever possible instead of specifying the string.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 3]
```diff
@@ -183,6 +186,32 @@ public synchronized void create(final NamespaceMeta metadata) throws Exception {
...
+        String kubeNamespace = namespace.getNamespace()
+                .toLowerCase()
+                .replace("_", "-")
+                .substring(0, Math.min(namespace.getNamespace().length(), 63));
+        if (kubeNamespace.startsWith("-")) {
+          kubeNamespace = kubeNamespace.substring(1);
+        }
+        if (kubeNamespace.endsWith("-")) {
+          kubeNamespace = kubeNamespace.substring(0, kubeNamespace.length() - 1);
+        }
```
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Do we need to handle the case where multiple cdap namespaces map to the same k8s namespace?
example - <63_char_prefix>-1, <63_char_prefix>-2

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 4]
```diff
@@ -280,6 +285,57 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+      hardLimitMap.put("limits.memory", new Quantity(kubeMemoryLimit));
+    }
+    V1ResourceQuota resourceQuota = new V1ResourceQuota().apiVersion("v1").kind("ResourceQuota");
+    resourceQuota.setMetadata(new V1ObjectMeta()
+            .name("cdap-resource-quota")
+            .putLabelsItem("cdapNamespace", cdapNamespace));
+    resourceQuota.setSpec(new V1ResourceQuotaSpec().hard(hardLimitMap));
+
+    try {
+      coreV1Api.createNamespace(namespace, null, null, null);
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: what happens if the namespace already exists? Is an exception thrown?
If yes, should we handle it differently depending on whether the k8s namespace was created by CDAP (some cleanup failed?) vs created manually by the user?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java (hunk 3 above)
Comment (nliu132) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: There's no character limit for CDAP namespaces, so that's a possible case. I also just thought of another edge case - `TEST` and `test` would be considered unique CDAP namespaces but would resolve to the same Kube namespace. 
I wonder if it would be better then to have the Kube namespace be some hash or ID generated from the CDAP name, instead of trying to keep the 2 as similar as possible

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 4 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: the `listNamespace` check in the beginning of the method throws an exception, regardless if it was created by CDAP or the user. I think we should handle it the same (i.e. disallow creation), since in the case where some CDAP cleanup failed, it's possible that the new namespace creation request has some different properties than the partially deleted/existing Kube resources (which I guess is only CPU/memory limit for now)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 5]
```diff
@@ -183,6 +186,32 @@ public synchronized void create(final NamespaceMeta metadata) throws Exception {
         storageProviderNamespaceAdmin.get().create(metadata);
         return null;
       });
+
+      // if needed, run master environment specific logic
+      MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+      if (masterEnv != null && masterEnv.getName().equals("k8s")) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Should have `"k8s"` checks here. All the k8s specific logic should be pushed inside the k8s master env

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 6]
```diff
@@ -369,6 +404,11 @@ public synchronized void updateProperties(NamespaceId namespaceId, NamespaceMeta
       throw new BadRequestException(String.format("Mappings %s for namespace %s cannot be updated once the namespace " +
                                                     "is created.", difference, namespaceId));
     }
+
+    if (namespaceMeta.getKubeNamespace() != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: We are leaking implementation details. There shouldn't be `kubeNamespace` in the NamespaceMeta

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 7]
```diff
@@ -121,6 +126,11 @@
   private static final String CDAP_LOCALIZE_FILES_PATH = "/etc/cdap/localizefiles";
   private static final String CDAP_CONFIG_MAP_PREFIX = "cdap-compressed-files-";
 
+  private static final String NAMESPACE_NAME_PROPERTY = "kubernetesNamespace";
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Please follow dot separate naming convention. Call it `k8s.namespace`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 8]
```diff
@@ -121,6 +126,11 @@
...
+  private static final String NAMESPACE_CPU_LIMIT_PROPERTY = "kube.limits.cpu";
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Please following existing naming convention. Call them `k8s.namespace.cpu.limits` and `k8s.namespace.memory.limits`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 9]
```diff
@@ -121,6 +126,11 @@
...
+  private static final String NAMESPACE_MEMORY_LIMIT_PROPERTY = "kube.limits.memory";
+  private static final String CDAP_NAMESPACE_LABEL = "cdapNamespace";
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Call it `cdap.namespace`

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 10]
```diff
@@ -280,6 +290,57 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
                            sparkConfMap);
   }
 
+  @Override
+  public void onNamespaceCreation(String cdapNamespace, Map<String, String> properties) throws Exception {
+    String kubeName = properties.get(NAMESPACE_NAME_PROPERTY);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Better call this variable `namespace` as it is the k8s namespace, not a name.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 11]
```diff
@@ -280,6 +290,57 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+    try {
+      V1NamespaceList namespaceList = coreV1Api.listNamespace(null, null, null,
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:310]: Why use the list API instead of the read API?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 12]
```diff
@@ -280,6 +290,57 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+              String.format("metadata.name=%s", kubeName), null, 1, null, null, 10, false);
+      if (!namespaceList.getItems().isEmpty()) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:312]: The k8s namespace can exist if it carries the cdap namespace annotation that matches with the `cdapNamespace`, as it can be from a previously attempt of creation (and failed), right?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 13]
```diff
@@ -553,4 +614,27 @@ private void populateLabels(Map<String, String> sparkConfMap) {
     }
     return sparkConfs;
   }
+
+  /**
+   * Deletes Kubernetes namespace and associated resources if they exist for the given CDAP namespace
+   */
+  private void deleteKubeNamespace(String cdapNamespace) throws Exception {
+    try {
+      V1NamespaceList namespaceList = coreV1Api.listNamespace(null, null, null, null,
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: It's better to always just issue an delete and catch the exception to check if the deletion failed due to not found (which is fine), or due to some other error (which needs to propagate).

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 11 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:310]: the read API returns a 404 error if the namespace doesn't exist. It could be used instead of the list API, but I thought it may be weird to throw an error if the API doesn't throw an error, and to check if the error code is 404 or something else (which would indicate an unexpected ApiException)

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 12 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:312]: If there's a previous failed creation attempt, there's the possibility that the user could modify the k8s name/CPU limit/memory limit, so the existing resources are out of sync with the user input. It could be supported, but we'd need to check that a namespace exists with the `cdap.namespace` label and that it matches the new user input, and then check if a `cdap-resource-quota` exists and that it matches the new user input

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 14]
```diff
@@ -183,6 +189,13 @@ public synchronized void create(final NamespaceMeta metadata) throws Exception {
         storageProviderNamespaceAdmin.get().create(metadata);
         return null;
       });
+
+      // if needed, run master environment specific logic
+      MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+      if (masterEnv != null) {
+        String configJson = GSON.toJson(metadata.getConfig());
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: Shouldn't GSON to and from to get the map. You can turn the `NamespaceConfig.getConfigs()` method to be public to retrieve the map.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java [hunk 15]
```diff
@@ -275,6 +288,13 @@ public synchronized void delete(@Name("namespaceId") final NamespaceId namespace
 
     LOG.info("Deleting namespace '{}'.", namespaceId);
     try {
+      // if needed, run master environment specific logic
+      MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+      if (masterEnv != null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/namespace/DefaultNamespaceAdmin.java]: We probably should check if it is not the DEFAULT namespace. See the logic below.

File: cdap-kubernetes/src/test/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironmentTest.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static io.cdap.cdap.master.environment.k8s.KubeMasterEnvironment.NAMESPACE_PROPERTY;
```
Comment (chtyim) [inline cdap-kubernetes/src/test/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironmentTest.java]: Don't use import static.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 17]
```diff
@@ -46,10 +50,13 @@
 import io.kubernetes.client.openapi.models.V1Pod;
 import io.kubernetes.client.openapi.models.V1PodSpec;
 import io.kubernetes.client.openapi.models.V1PodSpecBuilder;
+import io.kubernetes.client.openapi.models.V1ResourceQuota;
+import io.kubernetes.client.openapi.models.V1ResourceQuotaSpec;
 import io.kubernetes.client.openapi.models.V1Volume;
 import io.kubernetes.client.openapi.models.V1VolumeMount;
 import io.kubernetes.client.util.Config;
 import io.kubernetes.client.util.Yaml;
+import org.apache.commons.lang3.StringUtils;
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Don't use the apache common library. Just check if the string is not null and `!isEmpty()`.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 18]
```diff
@@ -279,6 +293,64 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
                            sparkConfMap);
   }
 
+  @Override
+  public void onNamespaceCreation(String cdapNamespace, Map<String, String> properties) throws Exception {
+    String namespace = properties.get(NAMESPACE_PROPERTY);
+    if (StringUtils.isEmpty(namespace)) {
+      LOG.info("Kubernetes namespace not created for {} because no name was provided in the config", cdapNamespace);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: How does it affect subsequence operations? Shouldn't this a required property in the k8s environment (if enabled)?

BTW, there should be a master env configuration to control if namespace creation is required or not (e.g. `master.environment.k8s.namespace.creation.enabled`)

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 12 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:312]: Yes. Maybe we should implement that logic after this PR.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 19]
```diff
@@ -279,6 +293,64 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+              + e.getCode() + ", Body = " + e.getResponseBody(), e);
+    }
+
+    V1Namespace namespaceObject = new V1Namespace();
+    namespaceObject.setMetadata(new V1ObjectMeta().name(namespace).putLabelsItem(CDAP_NAMESPACE_LABEL, cdapNamespace));
+
+    String kubeCpuLimit = properties.get(NAMESPACE_CPU_LIMIT_PROPERTY);
+    String kubeMemoryLimit = properties.get(NAMESPACE_MEMORY_LIMIT_PROPERTY);
+    Map<String, Quantity> hardLimitMap = new HashMap<>();
+    if (StringUtils.isNotEmpty(kubeCpuLimit)) {
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Use `Quantity.parse` instead. It handles `null` and empty string.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 20]
```diff
@@ -279,6 +293,64 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+    }
+    V1ResourceQuota resourceQuota = new V1ResourceQuota();
+    resourceQuota.setMetadata(new V1ObjectMeta()
+            .name("cdap-resource-quota")
+            .putLabelsItem(CDAP_NAMESPACE_LABEL, cdapNamespace));
+    resourceQuota.setSpec(new V1ResourceQuotaSpec().hard(hardLimitMap));
+
+    try {
+      coreV1Api.createNamespace(namespaceObject, null, null, null);
+      LOG.info("Created Kubernetes namespace {}", namespace);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Suggest to have the message as `Create Kubernetes namespace {} for namespace {}`.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 21]
```diff
@@ -279,6 +293,64 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+      if (!hardLimitMap.isEmpty()) {
+        coreV1Api.createNamespacedResourceQuota(namespace, resourceQuota, null, null, null);
+        LOG.info("Created Kubernetes resource quota {}", resourceQuota);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Similarly, log with the k8s namespace and the CDAP namespace in the message.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 22]
```diff
@@ -553,4 +625,28 @@ private void populateLabels(Map<String, String> sparkConfMap) {
...
   }
+
+  /**
+   * Deletes Kubernetes namespace and associated resources if they exist for the given CDAP namespace
+   */
+  private void deleteKubeNamespace(String namespace) throws Exception {
+    try {
+      // PropagationPolicy is set to background cascading deletion. Kubernetes deletes the owner object immediately and
+      // the controller cleans up the dependent objects in the background.
+      coreV1Api.deleteNamespace(namespace, null, null, 0, null, "Background", null);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:646]: Note for next PR. We should first do a get / list of the namespace, retrieve the metadata to make sure it is a namespace created by CDAP on the intended CDAP namespace before. Then do a conditional delete by matching with the resource version that you retrieved from the get / list call.

Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:646]: Does the ResourceQuota get deleted automatically when the namespace is deleted?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java [hunk 23]
```diff
@@ -48,13 +51,14 @@
...
   }
 
-  public NamespaceConfig(String schedulerQueueName, @Nullable String rootDirectory,
-                         @Nullable String hbaseNamespace, @Nullable String hiveDatabase,
-                         @Nullable String principal, @Nullable String groupName, @Nullable String keytabURI,
-                         boolean exploreAsPrincipal) {
+  public NamespaceConfig(String schedulerQueueName, @Nullable String rootDirectory, @Nullable String hbaseNamespace,
+                         @Nullable String hiveDatabase, @Nullable String principal, @Nullable String groupName,
+                         @Nullable String keytabURI, boolean exploreAsPrincipal, @Nullable String kubeNamespace,
+                         @Nullable String kubeCpuLimit, @Nullable String kubeMemoryLimit) {
```
Comment (chtyim) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java]: Shouldn't have the kube properties in the constructor. Use the constructor that takes a Map. Or make this constructor to take a `Map` in addition to those existing properties.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java [hunk 24]
```diff
@@ -40,6 +40,9 @@
   public static final String PRINCIPAL = "principal";
   public static final String GROUP_NAME = "groupName";
   public static final String KEYTAB_URI = "keytabURI";
+  public static final String KUBE_NAMESPACE = "k8s.namespace";
```
Comment (chtyim) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java]: Shouldn't have these constants.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceMeta.java [hunk 25]
```diff
@@ -88,6 +88,9 @@ public NamespaceConfig getConfig() {
     private int keytabURIVersion;
     private long generation = 0;
     private boolean exploreAsPrincipal = true;
+    private String kubeNamespace;
```
Comment (chtyim) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceMeta.java]: Shouldn't have k8s specific fields in this class.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 22 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java:646]: Yes, deleting a namespace will [delete its resources as well ](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace)

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 18 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: You're right - I was afraid this would interfere with other parallel Kubernetes related work, but the master env configuration should determine if this logic is required

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 19 above)
Comment (nliu132) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: `parse` will throw a `QuantityFormatException` if the value is null or empty. Since the properties are optional, I think it'd be better to do the null/empty check instead of catching and ignoring the exception

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 26]
```diff
@@ -120,6 +127,13 @@
   private static final String CDAP_LOCALIZE_FILES_PATH = "/etc/cdap/localizefiles";
   private static final String CDAP_CONFIG_MAP_PREFIX = "cdap-compressed-files-";
 
+  private static final String NAMESPACE_CREATION_ENABLED = "master.environment.k8s.namespace.creation.enabled";
+  @VisibleForTesting
+  static final String NAMESPACE_PROPERTY = "k8s.namespace";
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: It's better to group together private static and package static separately.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 27]
```diff
@@ -279,6 +295,68 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+  @Override
+  public void onNamespaceCreation(String cdapNamespace, Map<String, String> properties) throws Exception {
+    // check if namespace creation is enabled from master config
+    if (!namespaceCreationEnabled) {
+      return;
+    }
+    String namespace = properties.get(NAMESPACE_PROPERTY);
+    if (namespace == null || namespace.isEmpty()) {
+      throw new IOException(String.format("Cannot create Kubernetes namespace for %s because no name was provided",
+              cdapNamespace));
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Typically the newline is aligned with the beginning of the same scope in the previous line. E.g.

```
      throw new IOException(String.format("Cannot create Kubernetes namespace for %s because no name was provided",
                                          cdapNamespace));
```

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 28]
```diff
@@ -553,4 +631,33 @@ private void populateLabels(Map<String, String> sparkConfMap) {
...
+   */
+  private void deleteKubeNamespace(String namespace) throws Exception {
+    try {
+      // PropagationPolicy is set to background cascading deletion. Kubernetes deletes the owner object immediately and
+      // the controller cleans up the dependent objects in the background.
+      coreV1Api.deleteNamespace(namespace, null, null, 0, null, "Background", null);
+      LOG.info("Deleted Kubernetes namespace and associated resources for {}", namespace);
+    } catch (ApiException e) {
+      if (e.getCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+        LOG.info("Kubernetes namespace {} was not deleted because it was not found", namespace);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Log as debug.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 29]
```diff
@@ -279,6 +295,68 @@ public SparkConfig generateSparkSubmitConfig(SparkSubmitContext sparkSubmitConte
...
+
+    try {
+      coreV1Api.createNamespace(namespaceObject, null, null, null);
+      LOG.info("Created Kubernetes namespace {} for namespace {}", namespace, cdapNamespace);
+      if (!hardLimitMap.isEmpty()) {
+        coreV1Api.createNamespacedResourceQuota(namespace, resourceQuota, null, null, null);
+        LOG.info("Created resource quota for Kubernetes namespace {} for namespace {}", namespace, cdapNamespace);
+      }
+    } catch (ApiException e) {
+      deleteKubeNamespace(namespace);
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: It's better to catch the exception from the `deleteKubeNamespace` call and add it as a suppressed exception of the original exception (`e`) so that the root cause of the creation failure is not lost.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java [hunk 30]
```diff
@@ -55,7 +55,14 @@ public NamespaceConfig(String schedulerQueueName, @Nullable String rootDirectory
...
-    Map<String, String> configs = new HashMap<>();
+    this(schedulerQueueName, rootDirectory, hbaseNamespace, hiveDatabase, principal, groupName, keytabURI,
+            exploreAsPrincipal, new HashMap<>());
+  }
+
+  public NamespaceConfig(String schedulerQueueName, @Nullable String rootDirectory,
+                         @Nullable String hbaseNamespace, @Nullable String hiveDatabase,
+                         @Nullable String principal, @Nullable String groupName, @Nullable String keytabURI,
+                         boolean exploreAsPrincipal, Map<String, String> configs) {
     configs.put(SCHEDULER_QUEUE_NAME, schedulerQueueName);
```
Comment (chtyim) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/NamespaceConfig.java:67]: We shouldn't modify the incoming map. You need to copy it to to a local variable first.

```java
Map<String, String> localConfigs = new HashMap<>(configs);
localConfigs.put(...);
...
this.configs = Collections.unmodifiableMap(localConfigs);
```

**Review by chtyim — COMMENTED** [review]
Couple more minor comments.

**Review by chtyim — APPROVED** [review]
Please squash the commits before merging.
